File: L03/python/udp_client.py
#!/usr/bin/python3
#*
#******************************************************************************
#* @file    udp_client.py
#* @author  AW           Adrian.Wojcik@put.poznan.pl
#* @version 1.0
#* @date    29-Mar-2022
#* @brief   Simple UDP client
#* 
#******************************************************************************
#*
import socket
from datetimestr import datetimestr
from udp_parse_arg import parse_args

if __name__ == '__main__':
    
    port, ip = parse_args()

    # Set default values if port or ip are not specified
    if port is None:
        port = 20000
    if ip is None:
        ip = "172.27.229.191" # C / Python server [RPi]
        #ip = "192.168.0.15" # MATLAB server [PC]

    msgFromClient       = "Hello from [Python] client"
    bytesToSend         = str.encode(msgFromClient)
    serverAddressPort   = (ip, port)   # C / Python server [RPi]
    bufferSize          = 1024

    # Create a UDP socket at client side
    UDPClientSocket = socket.socket(family=socket.AF_INET, type=socket.SOCK_DGRAM)

    # Send to server using created UDP socket
    UDPClientSocket.sendto(bytesToSend, serverAddressPort)
    print("Message sent.")

    msgFromServer = UDPClientSocket.recvfrom(bufferSize)

    msg = "Server response [{}]: {}".format(datetimestr(), msgFromServer[0])
    print(msg)

File: L03/c++/udp_client.c
/**
  ******************************************************************************
  * @file    udp_client.cpp
  * @author  AW           Adrian.Wojcik@put.poznan.pl
  * @version 1.0
  * @date    29-Mar-2022
  * @brief   Simple UDP client
  *
  ******************************************************************************
  */

/* Includes ------------------------------------------------------------------*/
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <netinet/in.h>

#include "datetimestr.h"
#include "udp_parse_args.h"

/* Define --------------------------------------------------------------------*/
#define DEFAULT_IP		"172.27.229.191" // C / Python server [Linux server]
//#define DEFAULT_IP		"192.168.0.15" // MATLAB server [Windows PC]
#define DEFAULT_PORT	20000
#define MAXLINE  		1024

/* Main function -------------------------------------------------------------*/

/**
  * @brief  The application entry point.
  * @param[in] argc : argument count; number of command-line arguments passed 
  *                   by the user including the name of the program.
  * @param[in] argv : argument vector; character pointers (C-strings) listing 
  *                   all the arguments.
  * @retval 0 
  */
int main(int argc, char* argv[])
{
	struct argopt opt = { DEFAULT_PORT, DEFAULT_IP };
	parse_args(argc, argv, &opt);
	
	int sockfd;
	char buffer[MAXLINE] = {0,};
	char hello[] = "Hello from [C] client";
	struct sockaddr_in	 servaddr;

	// Creating socket file descriptor
	if ( (sockfd = socket(AF_INET, SOCK_DGRAM, 0)) < 0 ) {
		perror("socket creation failed");
		exit(EXIT_FAILURE);
	}

	memset(&servaddr, 0, sizeof(servaddr));
	
	// Filling server information
	servaddr.sin_family = AF_INET; // IPv4
	servaddr.sin_port = htons(opt.port);
	servaddr.sin_addr.s_addr = inet_addr(opt.ip);
	
	// Send to socket
	sendto(sockfd, (const char *)hello, strlen(hello), 0, (const struct sockaddr*)&servaddr, sizeof(servaddr));
	printf("Message sent.\n");
	
	// Receive from socket
	int len = sizeof(servaddr);
	recvfrom(sockfd, (char *)buffer, MAXLINE, 0, (struct sockaddr*)&servaddr, (socklen_t*)&len);
	printf("Server response [%s]: %s\n", datetimestr(), buffer);

	close(sockfd);
	return 0;
}
